cpp/src/arrow/util/binary_view_util.h

```diff
@@ -0,0 +1,163 @@
+  memcpy(&out.raw.prefix, data, sizeof(out.raw.prefix));
+  return out;
+}
+
+inline BinaryViewType::c_type ToIndexOffsetBinaryView(std::string_view v,
+                                                      int32_t buffer_index,
+                                                      int32_t offset) {
+  return ToIndexOffsetBinaryView(v.data(), static_cast<int32_t>(v.size()), buffer_index,
+                                 offset);
+}
+
+inline std::string_view FromRawPointerBinaryView(const BinaryViewType::c_type& v) {
+  auto* data = v.is_inline() ? v.inlined.data.data() : v.raw.data;
+  return {reinterpret_cast<const char*>(data), static_cast<size_t>(v.size())};
+}
+inline std::string_view FromRawPointerBinaryView(BinaryViewType::c_type&&) = delete;
+
+template <typename BufferPtr>
+std::string_view FromIndexOffsetBinaryView(const BinaryViewType::c_type& v,
+                                           const BufferPtr* data_buffers) {
+  auto* data = v.is_inline() ? v.inlined.data.data()
+                             : data_buffers[v.io.buffer_index]->data() + v.io.offset;
+  return {reinterpret_cast<const char*>(data), static_cast<size_t>(v.size())};
+}
+template <typename BufferPtr>
+std::string_view FromIndexOffsetBinaryView(BinaryViewType::c_type&&,
+                                           const BufferPtr*) = delete;
+
+inline int MemcmpRawPointerBinaryViewSkipPrefix(int32_t size, BinaryViewType::c_type l,
+                                                BinaryViewType::c_type r) {
+  return memcmp(l.raw.data + BinaryViewType::kPrefixSize,  //
+                r.raw.data + BinaryViewType::kPrefixSize,  //
+                size - BinaryViewType::kPrefixSize);
+}
+
+template <typename BufferPtr>
+inline int MemcmpIndexOffsetBinaryViewSkipPrefix(int32_t size, BinaryViewType::c_type l,
+                                                 BinaryViewType::c_type r,
+                                                 const BufferPtr* l_buffers,
+                                                 const BufferPtr* r_buffers) {
+  return memcmp(
+      l_buffers[l.io.buffer_index]->data() + l.io.offset + BinaryViewType::kPrefixSize,
+      r_buffers[r.io.buffer_index]->data() + r.io.offset + BinaryViewType::kPrefixSize,
+      size - BinaryViewType::kPrefixSize);
+}
+
+template <typename... BufferPtr>
+bool EqualBinaryViewImpl(BinaryViewType::c_type l, BinaryViewType::c_type r,
+                         const BufferPtr*... buffers) {
+  int64_t l_size_and_prefix, r_size_and_prefix;
+  memcpy(&l_size_and_prefix, &l, sizeof(l_size_and_prefix));
+  memcpy(&r_size_and_prefix, &r, sizeof(r_size_and_prefix));
+
+  if (l_size_and_prefix != r_size_and_prefix) return false;
+
+  if (l.is_inline()) {
+    // The inline part is zeroed at construction, so we can compare
+    // a word at a time if data extends past 'prefix_'.
+    int64_t l_inlined, r_inlined;
+    memcpy(&l_inlined,
+           AssumeAlignedAs<int64_t>(l.inline_data() + BinaryViewType::kPrefixSize),
```

Comment:
memcpy + AssumeAlignedAs gives the compiler an explicit hint that the call to memcpy can be replaced with an aligned load, which is the minimum cost access to these bytes which I can think of. Directly calling memcmp would *probably* be optimized down to the whole word comparison, but for an operation as performance sensitive as equality I think it's worthwhile to make the fastest path explicit. FWIW there are similar fast paths available for hashing and ordered comparison which aren't in this PR

Reply:
It's true for most compilers and architectures there's no difference: https://godbolt.org/z/oxdz3r4fj

I'll replace this with `memcmp` if you prefer. Alternatively, I can reduce the boilerplate here by adding `AlignedLoadAs<>()` to match `SafeLoadAs<>()`

Reply:
`SafeLoadAs`'s purpose is already to avoid UB on unaligned loads, so I'm not sure why an additional function would be needed.

Reply:
I can only speculate on that as I don't know *how much* existing code a hypothetical future optimization pass might break or what the pass' author's priorities would be. It could happen that arrow C++ is the largest project broken by the pass, and it could happen that they would add the pass to `-O2`. Even in that case as long as arrow is still being maintained the fix would probably not be difficult or the pass could be independently disabled. `</speculation>` It'd be my preference to err on the side of pedantia where language guarantees are concerned.

Reply:
just to chime in here, I personally would err on the side of avoiding UB when it comes to language spec guarantees. If the issue here is that this is "gratuitously pedantic" but removing the pedantic-ness would require effectively UB, i'd rather us be pedantic. I also don't think that this pedantic-ness should be something that blocks us from getting this merged.

Reply:
Urban legends about alignment tend to stick around and produce code that is more annoying to maintain.
That's why I want to make sure that the complication is necessary and not gratuitous.

Reply:
That's a fair point, I'd just rather "annoying to maintain but correct/no UB" vs "slightly nicer to maintain but relies on UB". I've learned over the years that every time I've used the phrase "I don't think any sane compiler would...." I've eventually been proven wrong by some version of some compiler that ends up being used, lol. 

Unless there's a better suggestion that doesn't rely on UB, my vote is to keep the gratuitousness